music_assistant/server/providers/siriusxm/__init__.py

```diff
@@ -0,0 +1,437 @@
+                        name=category.name,
+                        label="",
+                    )
+
+        if sub_path.startswith("categories/"):
+            category_key = sub_sub_path
+            channels = self._channels_by_category[category_key]
+            for channel in channels:
+                yield self._parse_radio(channel)
+
+    async def get_library_radios(self) -> AsyncGenerator[Radio, None]:
+        """Retrieve library/subscribed playlists from the provider."""
+        for channel_guid, channel in self._channels_by_id.items():
+            yield self._parse_radio(channel)
+
+    async def get_radio(self, channel_id: str) -> Union[Radio, None]:
+        channel = self._channels_by_id[channel_id] if channel_id in self._channels_by_id else None
+        if channel is None:
+            return None
+        return self._parse_radio(channel)
+
+    async def get_stream_details(self, item_id: str) -> StreamDetails:
+        """Get StreamDetails for a radio station."""
+        # channel_id = item_id
+        # channel = self._channels_by_id[channel_id]
+
+        # playlist = await self._client.get_playlist(channel.id, use_cache=False)
+        # playlist_items = list(filter(lambda l: not l.startswith("#"), playlist.split('\n')))
+
+        # first_segment = await self._client.get_segment(playlist_items[0])
+
+        # primary_hls_root = await self._client.get_primary_hls_root()
+        # first_item = f"{primary_hls_root}/{playlist_items[0]}"
+
+        # live_channel_raw = await self._client.get_now_playing(channel)
+        # live_channel = XMLiveChannel.from_dict(live_channel_raw["moduleList"]["modules"][0])
+
+        # playlist_url = live_channel.primary_hls.url
+
+        # cache_key = f"{self.instance_id}.media_info.{channel_id}"
+        # cached_info = await self.mass.cache.get(cache_key)
+        # if cached_info:
+        #     media_info = AudioTags.parse(cached_info)
+        # else:
+        #     # parse info with ffprobe (and store in cache)
+        #     media_info = await parse_tags(self._to_generator(first_segment))
+        #     await self.mass.cache.set(cache_key, media_info.raw)
+
+        # MPEG AAC Audio (mp4a)
+        # Type: Audio
+        # Channels: 2
+        # Sample Rate: 44100
+        # Bits per sample: 32
+
+        return StreamDetails(
+            provider=self.instance_id,
+            item_id=item_id,
+            audio_format=AudioFormat(
+                sample_rate=44100,
+                channels=2,
+                bit_depth=32,
+                content_type=ContentType.M4A,
+            ),
+            # stream_title=channel.name,
+            # duration=?,
+            # data={"url": url, "format": url_details["format"]}
+            # expires=time() + 24 * 3600,
+            # size=?,
+            media_type=MediaType.RADIO,
+            # data=live_channel.primary_hls.url,
+            direct=None,
+            # can_seek=False,

```

Comment:
can_seek should be set to False because these streams are not seekable

Reply:
I'll set that for now. Their streams do support some DVR-like transport options. You can seek backwards or forward if you're paused a bit (not the future obviously)... but I don't know if I can implement that or not with their client. The JSON payload has like cut lists that tell you when songs/episodes start & stop and you can skip around by those fine/course-grained options a bit. But there are per-channel DRM restrictions in the payload about the number of seconds you can skip by and the maximum number of skips allowed in a certain time window.

I'll go with `can_seek=False` for now. I think music assistant would probably need to extend the provider model a little bit to support these kinds of things.

Reply:
ooh fancy... yeah that would require extending our models a bit in the future yeah